### PR TITLE
Update hello-node example to use deployments

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: docs/hellonode.md [hunk 1]
```diff
@@ -83,14 +83,14 @@ Now there is a trusted source for getting an image of your containerized app.
 Let's try your image out with Docker:
 
 ```shell
-docker run -d -p 8080:8080 gcr.io/PROJECT_ID/hello-node:v1
+$ docker run -d -p 8080:8080 gcr.io/PROJECT_ID/hello-node:v1
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 2]
```diff
@@ -83,14 +83,14 @@ Now there is a trusted source for getting an image of your containerized app.
...
 325301e6b2bffd1d0049c621866831316d653c0b25a496d04ce0ec6854cb7998
 ```
 
 Visit your app in the browser, or use `curl` or `wget` if you’d like :
 
 ```shell
-curl http://localhost:8080
+$ curl http://localhost:8080
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 3]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
 Create a pod with the `kubectl run` command:
 
 ```shell
-kubectl run hello-node \
+$ kubectl run hello-node \
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 4]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+The flag used in this command specifies that we’ll be using the load-balancer provided by the underlying infrastructure (in this case the [Compute Engine load balancer](https://cloud.google.com/compute/docs/load-balancing/)). Note that we expose the deployment, and not the pod directly.  This will cause the resulting service to load balance traffic across all pods managed by the deployment (in this case only 1 pod, but we will add more replicas later).
 
 The Kubernetes master creates the load balancer and related Compute Engine forwarding rules, target pools, and firewall rules to make the service fully accessible from outside of Google Cloud Platform.
 
-To find the publicly-accessible IP address, ask `kubectl` to describe the `hello-node` cluster service:
+To find the ip addresses associated with the service.
 
 ```shell
-kubectl get services hello-node
+$ kubectl get services hello-node
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No $

File: docs/hellonode.md [hunk 5]
```diff
@@ -221,36 +261,28 @@ First, let’s modify the application. On the development machine, edit server.j
...
-docker build -t gcr.io/PROJECT_ID/hello-node:v2 . 
-docker push gcr.io/PROJECT_ID/hello-node:v2
+$ docker build -t gcr.io/PROJECT_ID/hello-node:v2 . 
+$ docker push gcr.io/PROJECT_ID/hello-node:v2
 ```
 
 Building and pushing this updated image should be much quicker as we take full advantage of the Docker cache.
 
-We’re now ready for kubernetes to smoothly update our replication controller to the new version of the application:
+We’re now ready for kubernetes to smoothly update our deployment to the new version of the application.  Change the image from gcr.io/PROJECT_ID/hello-node:v1 to gcr.io/PROJECT_ID/hello-node:v2 using `kubectl edit`:
```
Comment (johndmulhausen) [inline docs/hellonode.md]: It's a little cloak-and-dagger what's happening here. How does "edit" know that you're swapping image labels when they aren't mentioned in the example command?

File: docs/hellonode.md [hunk 6]
```diff
@@ -221,36 +261,28 @@ First, let’s modify the application. On the development machine, edit server.j
...
 
-3. The original replication controller is deleted.
+```
+$ kubectl get deployments
+NAME         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+hello-node   4         5         4            3           1h
+```
 
-While this is happening, the users of the services should not see any interruption. After a little while they will start accessing the new version of your application. You can find more details on rolling updates in [this documentation](https://cloud.google.com/container-engine/docs/rolling-updates).
+While this is happening, the users of the services should not see any interruption. After a little while they will start accessing the new version of your application. You can find more details in the [deployment documentation](/docs/user-guide/deployments.md).
```
Comment (johndmulhausen) [inline docs/hellonode.md]: Remove ".md" from the end of the URL here

File: docs/hellonode.md [hunk 7]
```diff
@@ -108,7 +108,7 @@ $ docker stop 2c66d0efcbd4
 Now that the image works as intended and is all tagged with your `PROJECT_ID`, we can push it to the [Google Container Registry](https://cloud.google.com/tools/container-registry/), a private repository for your Docker images accessible from every Google Cloud project (but also from outside Google Cloud Platform) :
 
 ```shell
-gcloud docker push gcr.io/PROJECT_ID/hello-node:v1
+$ gcloud docker push gcr.io/PROJECT_ID/hello-node:v1
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 8]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+deployment "hello-node" created
 ```
 
-Now is probably a good time to run through some of the following interesting kubectl commands (none of these will change the state of the cluster, full documentation is available [here](https://cloud.google.com/container-engine/docs/kubectl/)):
+As shown in the output, the `kubectl run` created a **deployment** object.  Deployments are the recommended way for managing creation and scaling of pods.  In this example, a new deployment manages a single pod replica running the *hello-node:v1* image.
+
+To view the deployment we just created:
+
+```shell
+$ kubectl get deployments
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 9]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+NAME         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+hello-node   1         1         1            1           3m
+```
+
+To view the pod created by the deployment:
 
 ```shell
 $ kubectl get pods
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 10]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
-$ kubectl cluster-info
-$ kubectl config view
+NAME                         READY     STATUS    RESTARTS   AGE
+hello-node-714049816-ztzrb   1/1       Running   0          6m
+```
+
+To view the stdout / stderr from a pod (hello-node image has no output, so logs will be empty in this case):
+
+```shell
+$ kubectl logs <POD-NAME>
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 11]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+```
+
+To view metadata about the cluster:
+
+```shell
+$ kubectl get cluster-info
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 12]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+```
+
+To view cluster events:
+
+```shell
 $ kubectl get events
```
Comment (johndmulhausen) [inline docs/hellonode.md]: ...heh, clearly you're not the only one using these.

File: docs/hellonode.md [hunk 13]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
 ```
 
+Full documentation for kubectl commands is available [here](https://cloud.google.com/container-engine/docs/kubectl/)):
+
 At this point you should have our container running under the control of Kubernetes but we still have to make it accessible to the outside world.
 
 ## Allow external traffic
 
-By default, the pod is only accessible by its internal IP within the Kubernetes cluster. In order to make the `hello-node` container accessible from outside the kubernetes virtual network, you have to expose the pod as a kubernetes service.
+By default, the pod is only accessible by its internal IP within the Kubernetes cluster. In order to make the `hello-node` container accessible from outside the kubernetes virtual network, you have to expose the pod as a kubernetes **service**.
```
Comment (johndmulhausen) [inline docs/hellonode.md]: Link to "service" page?

File: docs/hellonode.md [hunk 14]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
 
 From our development machine we can expose the pod with the `kubectl` expose command and the `--type="LoadBalancer"` flag which creates an external IP to accept traffic:
 
 ```shell
-kubectl expose rc hello-node --type="LoadBalancer"
+$ kubectl expose deployment hello-node --type="LoadBalancer"
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 15]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+The flag used in this command specifies that we’ll be using the load-balancer provided by the underlying infrastructure (in this case the [Compute Engine load balancer](https://cloud.google.com/compute/docs/load-balancing/)). Note that we expose the deployment, and not the pod directly.  This will cause the resulting service to load balance traffic across all pods managed by the deployment (in this case only 1 pod, but we will add more replicas later).
 
 The Kubernetes master creates the load balancer and related Compute Engine forwarding rules, target pools, and firewall rules to make the service fully accessible from outside of Google Cloud Platform.
 
-To find the publicly-accessible IP address, ask `kubectl` to describe the `hello-node` cluster service:
+To find the ip addresses associated with the service.
```
Comment (johndmulhausen) [inline docs/hellonode.md]: Not a complete sentence :)

File: docs/hellonode.md [hunk 16]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
-hello-node   10.3.246.12   23.251.159.72   8080/TCP   run=hello-node   53s
+hello-node   10.3.246.12                   8080/TCP   run=hello-node   23s
 ```
 
-Note there are 2 IP addresses listed, both serving port 8080. One is the internal IP that is only visible inside your cloud virtual network; the other is the external load-balanced IP. In this example, the external IP address is 23.251.159.72. Traffic to the load-balanced IP will be load balanced to the three nodes you provisioned when initially creating the cluster!
+The EXTERNAL_IP may take several minutes to become available and visible.  If the EXTERNAL_IP is missing, wait a few minutes and try again.
 
-You should now be able to reach the service by pointing your browser to this address: http://<EXTERNAL_IP>**:8080**
+```shell
+$ kubectl get services hello-node
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 17]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+NAME         CLUSTER_IP    EXTERNAL_IP     PORT(S)    SELECTOR         AGE
+hello-node   10.3.246.12   23.251.159.72   8080/TCP   run=hello-node   2m
+```
+
+Note there are 2 IP addresses listed, both serving port 8080.  CLUSTER_IP is only visible inside your cloud virtual network.  EXTERNAL_IP is externally accessible.  In this example, the external IP address is 23.251.159.72.
```
Comment (johndmulhausen) [inline docs/hellonode.md]: Wrap the constants in backtics like `this`

File: docs/hellonode.md [hunk 18]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
+One of the powerful features offered by Kubernetes is how easy it is to scale your application. Suppose you suddenly need more capacity for your application; you can simply tell the deployment to manage a new number of replicas for your pod:
 
 ```shell
-kubectl scale rc hello-node --replicas=4
-$ kubectl get pods
-NAME               READY     STATUS    RESTARTS   AGE
-hello-node-6uzt8   1/1       Running   0          8m
-hello-node-gxhty   1/1       Running   0          34s
-hello-node-z2odh   1/1       Running   0          34s
+$ kubectl scale deployment hello-node --replicas=4
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 19]
```diff
@@ -133,74 +133,114 @@ A kubernetes **pod** is a group of containers, tied together for the purposes of
...
 ```
 
-You now have four replicas of your application, each running independently on the cluster with the load balancer you created earlier and serving traffic to all of them. 
+You now have four replicas of your application, each running independently on the cluster with the load balancer you created earlier and serving traffic to all of them.
 
 ```shell
-kubectl get rc hello-node
-CONTROLLER   CONTAINER(S)   IMAGE(S)                    SELECTOR         REPLICAS
-hello-node   hello-node     gcr.io/..../hello-node:v1   run=hello-node   3
+$ kubectl get deployment
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 20]
```diff
@@ -221,36 +261,28 @@ First, let’s modify the application. On the development machine, edit server.j
...
-    --image=gcr.io/PROJECT_ID/hello-node:v2 \
-    --update-period=2s
-Creating hello-node-324d23dd3e0e2474d6b76dc599abb519
-At beginning of loop: hello-node replicas: 2, hello-node-324d23dd3e0e2474d6b76dc599abb519 replicas: 1
-...
-At end of loop: hello-node replicas: 0, hello-node-324d23dd3e0e2474d6b76dc599abb519 replicas: 3
-Update succeeded. Deleting old controller: hello-node
-Renaming hello-node-324d23dd3e0e2474d6b76dc599abb519 to hello-node
-hello-node
+$ kubectl edit deployment hello-node
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

File: docs/hellonode.md [hunk 21]
```diff
@@ -221,36 +261,28 @@ First, let’s modify the application. On the development machine, edit server.j
...
-You should see in the standard output how the rolling update actually works:
-
-1. A new replication controller is created based on the new image
-
-2. The replica count on the new and old controllers is increased/decreased by one respectively until the desired number of replicas is reached
+This updates the deployment with the new image, causing new pods to be created with the new image and old pods to be deleted.
 
-3. The original replication controller is deleted.
+```
+$ kubectl get deployments
```
Comment (johndmulhausen) [inline docs/hellonode.md]: No "$"s, they make copy/pasting difficult.

**Comment by pwittrock** [comment]
@johndmulhausen RE: using $ in preformatted console, Happy to remove it from all the commands here, but I'd like to come up with a consistent way of clearly delineating the command from the command output (outside the scope of this PR).  I have come across an example or 2 where it wasn't clear and I ended up trying to run the command output in the console.  The [vagrant guide](http://kubernetes.io/docs/getting-started-guides/vagrant/) uses a couple different approaches to address this, which offers a side-by-side comparison.

File: docs/hellonode.md (hunk 17 above)
Comment (pwittrock) [inline docs/hellonode.md]: Done

File: docs/hellonode.md (hunk 15 above)
Comment (pwittrock) [inline docs/hellonode.md]: Done

**Comment by cmpis** [comment]
I tend to agree with @pwittrock on the `$`s. There are several cases that it was unclear if it was a command I could run or not.

**Comment by bgrant0607** [comment]
cc @janetkuo

File: docs/hellonode.md (hunk 13 above)
Comment (pwittrock) [inline docs/hellonode.md]: Done

File: docs/hellonode.md (hunk 6 above)
Comment (pwittrock) [inline docs/hellonode.md]: Done

File: docs/hellonode.md (hunk 13 above)
Comment (pwittrock) [inline docs/hellonode.md]: I've addressed this by linking the first reference to the docs.  Updated references for pod and deployment to link as well.

**Comment by janetkuo** [comment]
+1 for using `$`s.

File: docs/hellonode.md (hunk 3 above)
Comment (janetkuo) [inline docs/hellonode.md]: Suggest turn this to a single line.

File: docs/hellonode.md (hunk 11 above)
Comment (janetkuo) [inline docs/hellonode.md]: `$ kubectl get cluster-info` --> `$ kubectl cluster-info`.

Comment (janetkuo) [inline docs/hellonode.md]: `$ kubectl config view` is removed. Is this change intentional?

File: docs/hellonode.md (hunk 5 above)
Comment (pwittrock) [inline docs/hellonode.md]: Let me know if the new explanation is more clear.  Unfortunately the `kubectl edit` command forces the user to deal with the full deployment yaml config.  Let me know if you think there is a better way of tackling this or keeping it simple.

File: docs/hellonode.md (hunk 11 above)
Comment (janetkuo) [inline docs/hellonode.md]: The link to https://cloud.google.com/container-engine/docs/kubectl/ (which redirects to http://kubernetes.io/docs/user-guide/kubectl-overview/) is gone too.

File: docs/hellonode.md (hunk 5 above)
Comment (bgrant0607) [inline docs/hellonode.md]: Patch? Hopefully we'll have `kubectl set image` in 1.3.

File: docs/hellonode.md (hunk 11 above)
Comment (pwittrock) [inline docs/hellonode.md]: Good catch on the get cluster-info :)  I actually just caught it myself.  I can add kubectl config view back in.  I honestly don't remember if it was intentional or just an omission.  I feel like we are throwing a lot at the user in this section without a lot of explanation, and am not totally thrilled with it.

File: docs/hellonode.md (hunk 5 above)
Comment (pwittrock) [inline docs/hellonode.md]: Patch is a reasonable alternative.  Let me know if you think this is preferable to my recent update.  There is a trade off here between presenting the user with what looks like black magic vs presenting them with more information than necessary to accomplish the task at hand.  I am open to suggestions on the right balance to strike.

**Comment by pwittrock** [comment]
RE $: here is an example in the doc that I think may be confusing to the reader.  I find that it runs together a bit.  I would love to be able to shade command output differently with a ```shell-output``` directive or something.  Not necessary to solve as part of this PR.

```shell
gsutil ls # This is a command
gs://artifacts.<PROJECT_ID>.appspot.com/ # This is command output
gsutil rm -r gs://artifacts.<PROJECT_ID>.appspot.com/ # This is a command
Removing gs://artifacts.<PROJECT_ID>.appspot.com/... # This is command output
```

**Comment by pwittrock** [comment]
PTAL.  I believe I have addressed all comments.

Published at: http://pwittrock.github.io/docs/hellonode/

**Comment by pwittrock** [comment]
Realized http://pwittrock.github.io/docs/hellonode/ didn't have all of the updates, just pushed it.

**Comment by pwittrock** [comment]
Per an offline discussion, there are a few other changes that should be made either in this PR or as a follow up.

- Add documentation explaining that this requires both client and server version of 1.2 with link to 1.1 branch documentation
- Add documentation about how to get the old behavior (use generator)
- Explain how to setup kubectl
- Update the ui dashboard section to match what is shown in 1.2

File: docs/hellonode.md [hunk 22]
```diff
@@ -128,77 +128,120 @@ It’s now time to deploy your own containerized application to the Kubernetes c
...
 NAME         CLUSTER_IP    EXTERNAL_IP     PORT(S)    SELECTOR         AGE
-hello-node   10.3.246.12   23.251.159.72   8080/TCP   run=hello-node   53s
+hello-node   10.3.246.12   23.251.159.72   8080/TCP   run=hello-node   2m
 ```
 
-Note there are 2 IP addresses listed, both serving port 8080. One is the internal IP that is only visible inside your cloud virtual network; the other is the external load-balanced IP. In this example, the external IP address is 23.251.159.72. Traffic to the load-balanced IP will be load balanced to the three nodes you provisioned when initially creating the cluster!
+Note there are 2 IP addresses listed, both serving port 8080.  `CLUSTER_IP` is only visible inside your cloud virtual network.  `EXTERNAL_IP` is externally accessible.  In this example, the external IP address is 23.251.159.72.
 
-You should now be able to reach the service by pointing your browser to this address: http://<EXTERNAL_IP>**:8080**
+You should now be able to reach the service by pointing your browser to this address: http://<EXTERNAL_IP>**:8080** or running `curl http://<EXTERNAL_IP>:8080`
```
Comment (janetkuo) [inline docs/hellonode.md]: The first `<EXTERNAL_IP>` was escaped and `** **` wasn't rendered as bold text on the website. 
screen shot 2016-03-24 at 2 18 20 pm

File: docs/hellonode.md [hunk 23]
```diff
@@ -227,30 +270,82 @@ docker push gcr.io/PROJECT_ID/hello-node:v2
 
 Building and pushing this updated image should be much quicker as we take full advantage of the Docker cache.
 
-We’re now ready for kubernetes to smoothly update our replication controller to the new version of the application:
+We’re now ready for kubernetes to smoothly update our deployment to the new version of the application.  In order to change
+the image label for our running container, we will need to edit the existing *hello-node deployment* and change the image from
```
Comment (janetkuo) [inline docs/hellonode.md]: "Find the image field under `.spec.template.spec.containers` and change it from..."

File: docs/hellonode.md (hunk 22 above)
Comment (pwittrock) [inline docs/hellonode.md]: Fixed this issue, but worth understanding if it is expected for text in <> to disappear.  Is this because it is interpreted as an html tag?

Comment (janetkuo) [inline docs/hellonode.md]: Yes. http://daringfireball.net/projects/markdown/syntax#html
For example, `<b>abc</b>` gives you <b>abc</b>

**Comment by pwittrock** [comment]
The images also reference "Replication Controller" in the diagram  I am going to see if I can fix this.

**Comment by pwittrock** [comment]
Updated the image

File: docs/hellonode.md [hunk 24]
```diff
@@ -128,77 +128,120 @@ It’s now time to deploy your own containerized application to the Kubernetes c
...
 ```shell
-$ kubectl get pods
-$ kubectl logs
-$ kubectl cluster-info
-$ kubectl config view
-$ kubectl get events
+kubectl config view
 ```
 
+Full documentation for kubectl commands is available **[here](https://cloud.google.com/container-engine/docs/kubectl/)**:
```
Comment (johndmulhausen) [inline docs/hellonode.md]: This URL should be /docs/user-guide/kubectl-overview/

Comment (pwittrock) [inline docs/hellonode.md]: Done

**Comment by pwittrock** [comment]
RE: The ui

The old kubernetes-ui seems to be absent, and the new kubernetes-dashboard and unless I am missing something it looks like it still speaks replication controller.  This seems out of place given the example now just deals with replication controllers.  I am thinking we should simply omit it from this example.  Open to feedback on how we should generally reconcile the new dashboard with our deployment centric world.

File: docs/hellonode.md (hunk 22 above)
Comment (pwittrock) [inline docs/hellonode.md]: Good to know.  The code sections seem to ignore this.

**Comment by pwittrock** [comment]
PTAL

**Comment by janetkuo** [comment]
LGTM. 

I still prefer using `$`. [GKE docs](https://cloud.google.com/container-engine/docs/before-you-begin) and [Docker docs](https://docs.docker.com/engine/quickstart/) and most of our docs use `$`. it's more clear and consistent to our users IMHO.

**Comment by janetkuo** [comment]
>Open to feedback on how we should generally reconcile the new dashboard with our deployment centric world.

cc @bryk

**Comment by pwittrock** [comment]
I found a style guide that suggests using `$` when the code section includes command output, and omitting it if there is no command output.  I am happy to change it in a follow up if we agree on a standard.

**Comment by bryk** [comment]
> Open to feedback on how we should generally reconcile the new dashboard with our deployment centric world.

We'll be adding deployments, replica sets, and all the new stuff to the UI. We just didn't have time for this so far. 

Until this happens, I'm for removing the UI part. 

cc @kubernetes/dashboard-maintainers

**Comment by pwittrock** [comment]
SGTM, I will take care of that in an immediate follow up.

**Comment by pwittrock** [comment]
@bryk I can also add a dashboard section to hello-node once it has deployments + replicasets.

**Comment by johndmulhausen** [comment]
Thanks a ton for addressing the comments and chatting privately. :)
